### PR TITLE
[SVE2.1][Clang][LLVM]Add BFloat16 builtin in Clang and LLVM intrinisc

### PR DESCRIPTION
**Pull request by CarolineConcatto**
This patch implements the builtins in Clang
and the LLVM-IR intrinsic for the following:

For BFADD , BFSUB, BFMAX, BFMIN, BFMAXNM and BFMINNM, for instance: 
svbfloat16_t svadd[_bf16]_m (svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svadd[_bf16]_x (svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svadd[_bf16]_z (svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svadd[_n_bf16]_m (svbool_t pg, svbfloat16_t zdn, bfloat16_t zm); 
svbfloat16_t svadd[_n_bf16]_x (svbool_t pg, svbfloat16_t zdn, bfloat16_t zm); 
svbfloat16_t svadd[_n_bf16]_z (svbool_t pg, svbfloat16_t zdn, bfloat16_t zm); 
the add, could be replaced by sub, max, min, maxnm and minnm.

For BFMUL:
svbfloat16_t svmul[_bf16]_m(svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svmul[_bf16]_x(svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svmul[_bf16]_z(svbool_t pg, svbfloat16_t zdn, svbfloat16_t zm); 
svbfloat16_t svmul[_n_bf16]_m(svbool_t pg, svbfloat16_t zdn, bfloat16_t zm); 
svbfloat16_t svmul[_n_bf16]_x(svbool_t pg, svbfloat16_t zdn, bfloat16_t zm); 
svbfloat16_t svmul[_n_bf16]_z(svbool_t pg, svbfloat16_t zdn, bfloat16_t zm);

svbfloat16_t svmul_lane[_bf16](svbfloat16_t zn, svbfloat16_t zm,
                               uint64_t imm_idx);

For BFCLAMP:
svbfloat16_t svclamp[_bf16](svbfloat16_t op, svbfloat16_t min, svbfloat16_t max);

For BFMLA and BFMLS
svbfloat16_t svmla[_bf16]_m(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                            svbfloat16_t zm);
svbfloat16_t svmla[_bf16]_z(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                            svbfloat16_t zm);
svbfloat16_t svmla[_bf16]_x(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                            svbfloat16_t zm);
svbfloat16_t svmla[_n_bf16]_m(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                              bfloat16_t zm);
svbfloat16_t svmla[_n_bf16]_z(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                              bfloat16_t zm);
svbfloat16_t svmla[_n_bf16]_x(svbool_t pg, svbfloat16_t zda, svbfloat16_t zn,
                              bfloat16_t zm);

svbfloat16_t svmla_lane[_bf16](svbfloat16_t zda, svbfloat16_t zn,
                               svbfloat16_t zm, uint64_t

According to the PR#257[1]
[1]ARM-software/acle#257

Co-authored-by: Matthew Devereau <matthew.devereau@arm.com>

### CODE REVIEW

**Comment by momchil-velikov** [comment]
> Co-author: Matthew Devereau

Should be "Co-authored-by:"

File: clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmul_lane.c [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+// CHECK-NEXT:    ret <vscale x 8 x bfloat> [[TMP0]]
+//
+// CPP-CHECK-LABEL: @_Z25test_svmul_lane_bf16_idx7u14__SVBFloat16_tu14__SVBFloat16_t(
+// CPP-CHECK-NEXT:  entry:
+// CPP-CHECK-NEXT:    [[TMP0:%.*]] = tail call <vscale x 8 x bfloat> @llvm.aarch64.sve.fmul.lane.nxv8bf16(<vscale x 8 x bfloat> [[OP1:%.*]], <vscale x 8 x bfloat> [[OP2:%.*]], i32 7)
+// CPP-CHECK-NEXT:    ret <vscale x 8 x bfloat> [[TMP0]]
+//
+svbfloat16_t test_svmul_lane_bf16_idx7(svbfloat16_t op1, svbfloat16_t op2)
+{
+  return SVE_ACLE_FUNC(svmul_lane, _bf16, )(op1, op2, 7);
```
[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmul_lane.c:59]: This is needs to be a constant. If you check the arm_sve_sme.td you will see that this is the definitions of these intrinsics:
ddddi - {d} stands for default and {i} for  constant uint64_t.
The index lane always needs to be a constant.

[the remark this answers is not in the document]
Comment (momchil-velikov) [reply clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmul_lane.c:59]: Yes, that's what these tests do, check that you get an error if you pass a non-constant.

[the remark this answers is not in the document]
Comment (CarolineConcatto) [reply clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmul_lane.c:59]: Thank you Momchil,
Negative tests were added for bmful and bfmla and bfmla

**Comment by CarolineConcatto** [comment]
@momchil-velikov is the changes good?

**Review by momchil-velikov — APPROVED** [review]
LGTM, cheers!

Check the commit message, a double colon in `Co-authored-by::`.

**Comment by omjavaid** [comment]
This breaks on Arm/AArch64 lldb bots with llvm-mc not accepting sve2p1 when --mattr=all is supplied on commandline see error below:
`home/omair.javaid/work/llvm-dev/build/bin/llvm-mc -filetype=obj -triple aarch64-linux-gnueabihf /home/omair.javaid/work/llvm-dev/llvm-project/lldb/test/Shell/Commands/command-disassemble-aarch64-color.s -o /home/omair.javaid/work/llvm-dev/build/tools/lldb/test/Shell/Commands/Output/command-disassemble-aarch64-color.s.tmp --mattr=+all
/home/omair.javaid/work/llvm-dev/llvm-project/lldb/test/Shell/Commands/command-disassemble-aarch64-color.s:14:3: error: instruction requires: sve2p1
  bfadd z23.h, p3/m, z23.h, z13.h       // AEK_B16B16`

Following buildbots break:

https://lab.llvm.org/buildbot/#/builders/96
https://lab.llvm.org/buildbot/#/builders/17
https://lab.llvm.org/buildbot/#/builders/219

I am going to do a revert on this patch as this has been failing for 4 hours without a fix.

**Pull request merged**